Framework/Core/include/Framework/GroupedCombinations.h

```diff
@@ -0,0 +1,300 @@
+    {
+      if (!this->mIsEnd) {
+        this->addOne();
+        setCurrentGroupedCombination();
+      }
+      return *this;
+    }
+    // postfix increment
+    GroupedIterator operator++(int /*unused*/)
+    {
+      GroupedIterator copy(*this);
+      operator++();
+      return copy;
+    }
+    // return reference
+    reference operator*()
+    {
+      return *mCurrentGrouped;
+    }
+    bool operator==(const GroupedIterator& rh)
+    {
+      return (this->mIsEnd && rh.mIsEnd) || (this->mCurrent == rh.mCurrent);
+    }
+    bool operator!=(const GroupedIterator& rh)
+    {
+      return !(*this == rh);
+    }
+
+   private:
+    std::tuple<As...> getAssociatedTables()
+    {
+      auto& currentGrouping = GroupingPolicy::mCurrent;
+      constexpr auto k = sizeof...(As);
+      auto slicerIterators = functionToTuple<k>(&GroupSlicer<G, Us...>::begin, *mSlicer);
+      o2::soa::for_<k>([&](auto i) {
+        auto col = std::get<i.value>(currentGrouping);
+        for (auto& slice : *mSlicer) {
+          if (slice.groupingElement().globalIndex() == col.globalIndex()) {
+            std::get<i.value>(slicerIterators) = slice;
+            break;
+          }
+        }
+      });
+
+      return getSlices(slicerIterators, std::index_sequence_for<As...>());
+    }
+
+    template <std::size_t I, typename T, typename... Ts>
+    auto getSliceAt(std::tuple<Ts...>& t)
+    {
+      auto it = std::get<I>(t); // Get the tables corresponding to the grouping at index I
+      auto associatedType = it.template prepareArgument<T>();
+      return associatedType;
+    }
+
+    template <typename... Ts, std::size_t... Is>
+    std::tuple<As...> getSlices(std::tuple<Ts...>& t, std::index_sequence<Is...> is)
+    {
+      return std::make_tuple(getSliceAt<Is, As, Ts...>(t)...);
+    }
+
+    void setCurrentGroupedCombination()
+    {
+      std::tuple<As...> initAssociatedTables = getAssociatedTables();
+      constexpr auto k = sizeof...(As);
+      bool moveForward = false;
+      o2::soa::for_<k>([&](auto i) {
+        if (std::get<i.value>(initAssociatedTables).size() == 0) {
+          moveForward = true;
+        }
+      });
+      while (!this->mIsEnd && moveForward) {
+        GroupingPolicy::addOne();
+        std::tuple<As...> temp = getAssociatedTables();
+        moveForward = false;
+        o2::soa::for_<k>([&](auto i) {
+          if (std::get<i.value>(temp).size() == 0) {
+            moveForward = true;
+          }
+        });
+      }
+      std::tuple<As...> associatedTables = getAssociatedTables();
+
+      if (!this->mIsEnd) {
+        auto& currentGrouping = GroupingPolicy::mCurrent;
+        o2::soa::for_<k>([&](auto i) {
+          std::get<i.value>(associatedTables).bindExternalIndices(mGrouping.get());
+        });
+
+        mCurrentGrouped.emplace(interleaveTuples(currentGrouping, associatedTables));
+      }
+    }
+
+    std::shared_ptr<GroupSlicer<G, Us...>> mSlicer = nullptr;
+    std::shared_ptr<G> mGrouping;
```

Comment:
Why a shared ptr?

Reply:
With unique_ptr I cannot use the default copy constructor in l. 112, used e.g. in l. 244 - 259